### DISCUSSION TITLE
Polling for status

### DISCUSSION THREAD

**Discussion by splusq** [AnWeber/httpyac#589] [General]
I want to make a http request and poll (or loop in some way) until the response (a `status` field within the response) goes into completed state. Is this doable easily? Once completed, then run the asserts.

**Comment by AnWeber** [comment]
weird use case, but yes it works
```js
@shouldLoop={{true}}

# @loop while shouldLoop
GET https://httpbin.org/anything?foo={{$randomInt 1 9}}


{{@response
  async function writeStream(){
    const randomVal = response.parsedBody.args.foo;
    if(randomVal && +randomVal > 7){ // replace with response.statusCode === 200
      exports.shouldLoop= false;
      // asserts
    }
    await sleep(1000);
  }
  exports.waitPromise = writeStream();
}}
```

**Reply by splusq** [reply]
Thank you-your solution works. The only issue is the global variable, which has side-effects. 

What happens if one uses:

```
@ref foo
@loop foo.bar
``` 

when evaluating the loop is the `ref` re-run? It does not look like it - even if we use `@forceRef` it's not re-run should it be?

**Reply by AnWeber** [reply]
The variable is not global in the true sense of the word. It only exists in the context of this request. To simplify local work, variables that have already been defined are shared in the same file. However, this can be prevented using setting `httpyac.useRegionScopedVariables`.
But to the actual question. The `# @loop` statement only iterates over the area after the statement. Previous statements run outside the loop, i.e. the `# @ref` is only executed once.
Perhaps we should change the approach. In script areas it is also possible to trigger the execution of a `# @ref` yourself. Apparently the documentation is still missing, but you can access this [object](https://github.com/AnWeber/httpyac/blob/c7bc1c463657e7cfa9123366272cc907ad5ae10b/src/plugins/javascript/httpyacJsApi.ts#L22) and write a recursive function that calls itself up to the desired result.

```
GET https://httpbin.org/anything?foo={{$randomInt 1 9}}


{{@response
  async function recursiveLoop(){
    const randomVal = response.parsedBody.args.foo;
    if(randomVal && +randomVal > 5){ // replace with response.statusCode === 200
      return response;
      // asserts
    }else{
      return (await $httpyac.execute(httpRegion))?.foo;
    }
  }
  exports.foo = recursiveLoop();
}}
```
There would also be other solutions. A plugin/ interceptor for the execution, which sets the execution back to the beginning similar the index modification in the [loopMetaDataHandler](https://github.com/AnWeber/httpyac/blob/c7bc1c463657e7cfa9123366272cc907ad5ae10b/src/plugins/core/metaData/loopMetaDataHandler.ts#L70)

**Reply by splusq** [reply]
Thank you this is perfect! And thanks again for your amazing work!
